keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1883,79 +1868,214 @@ Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autosca
+// - at program startup:
+//   - instantiate plugin
+//   - EventsToRegister
+// - for each new pod:
+//   - PreEnqueue
+// - for each pod that is ready to be scheduled, one pod at a time:
+//   - PreFilter, Filter, etc.
+//
+// Cluster autoscaler works a bit differently. It identifies all pending pods,
+// takes a snapshot of the current cluster state, and then simulates the effect
+// of scheduling those pods with additional nodes added to the cluster. To
+// determine whether a pod fits into one of these simulated nodes, it
+// uses the same PreFilter and Filter plugins as the scheduler. Other extension
+// points (Reserve, Bind) are not used. Plugins which modify the cluster state
+// therefore need a different way of recording the result of scheduling
+// a pod onto a node. This is done through ClusterAutoScalerPlugin.
+//
+// Cluster autoscaler will:
+// - at program startup:
+//   - instantiate plugin, with real informer factory and no Kubernetes client
+//   - start informers
+// - at the start of a simulation:
+//   - call StartSimulation with a clean cycle state
+// - for each pending pod:
+//   - call PreFilter and Filter with the same cycle state that
+//     was passed to StartSimulation
+//   - call SimulateBindPod with the same cycle state that
+//     was passed to StartSimulation (i.e. *not* the one which was modified
+//     by PreFilter or Filter) to indicate that a pod is being scheduled onto a node
+//     as part of the simulation
+//
+// A plugin may:
+// - Take a snapshot of all relevant cluster state as part of StartSimulation
+//   and store it in the cycle state. This signals to the other extension
+//   points that the plugin is being used as part of the cluster autoscaler.
+// . In PreFilter and Filter use the cluster snapshot to make decisions
+//   instead of the normal "live" cluster state.
+// - In SimulateBindPod update the snapshot in the cycle state.
+type ClusterAutoScalerPlugin interface {
+	Plugin
+	// StartSimulation is called when the cluster autoscaler begins
+	// a simulation.
+	StartSimulation(ctx context.Context, state *CycleState) *Status
+	// SimulateBindPod is called when the cluster autoscaler decided to schedule
+	// a pod onto a certain node.
+	SimulateBindPod(ctx context.Context, state *CycleState, pod *v1.Pod, nodeInfo *NodeInfo) *Status
+	// SimulateEvictPod is called when the cluster autoscaler simulates removal
+	// of a node. All claims used only by this pod should be considered deallocated,
+	// to enable starting the same pod elsewhere.
+	SimulateEvictPod(ctx context.Context, state *CycleState, pod *v1.Pod, nodeName string) *Status
+    // NodeIsReady checks whether some real node has been initialized completely.
+    // Even if it is "ready" as far Kubernetes is concerned, some DaemonSet pod
+    // might still be missing or not done with its startup yet.
+    NodeIsReady(ctx context.Context, node *v1.Node) (bool, error)
```

Comment:
@MaciekPytel shouldn't this by done by cluster-autoscaler, outside of any extension?

Reply:
How would CA know which daemon set pods are needed and whether they have finished their startup work so that claims can be allocated for the node? It can't.

There are some hacks specifically for GPUs with extended resources in the current code base for this, but that won't work for DRA drivers.

Reply:
IIUC, all that CA needs to know is whether the Node will be able to eventually accommodate the Pod. So here it needs to know which Daemonsets Pods will be required if they aren't already there. This sounds independent of the driver.
In fact, knowing that a Node is not ready (right now) could potentially cause unnecessary scale ups.

Reply:
This is solved for GPU by knowing that a GPU node will always have a pre-defined (provider-specific) label and a gpu capacity. Autoscaler knows that a node that has the label, but no GPU capacity is still booting up. 
Another popular solution for this problem is using a 'startup-taint' - a node starts with a taint and the daemonset removes the taint once it's done with whatever intialization it needs to do. Startup taints are passed to CA via a flag or matched via regexp.

Finally, there is an [existing extension point](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/processors/customresources/custom_resources_processor.go) within autoscaler that can be used to implement this based on any other criteria (GPU and taints use this extension point internally).

Reply:
With DRA, a simple "capacity" is not enough for GPUs. The API for requesting GPUs is much more complex:
https://github.com/NVIDIA/k8s-dra-driver/blob/main/api/nvidia.com/resource/gpu/v1alpha1/gpuclaim.go#L26

Reply:
It doesn't need to be simple capacity for the model to work. The way it works today is actually pretty much the same as what is proposed here. The extension point I mentioned requires implementing `FilterOutNodesWithUnreadyResources` function which boils down to being able to tell if a node is fully initialized or not. And the taint-based solution can work with any initialization, it only requires that the node is started with a taint and the process doing the initialization removes it at the end.

We can add other mechanisms if needed. The only real requirement is that there are some clear, predefined criteria that let autoscaler tell if a node is ready or not based just on looking on the node object.